pkgs/top-level/all-packages.nix

```diff
@@ -355,6 +355,10 @@ with pkgs;
 
   iconConvTools = callPackage ../build-support/icon-conv-tools {};
 
+  ### OVERLAYS (n.b.: don't add to these without consulting the community; they can be expensive)
+
+  headless-overlay = import ./headless.nix;
+  headless = lib.lowPrioSet (recurseIntoAttrs (import path { overlays = [ headless-overlay ]; }));
```

Comment:
Won't this disregard any overlays and config the user chose to apply? Would it be possible to determine what Nixpkgs was initially called with, and use that here?

Reply:
I could probably merge it with the parameter set passed into nixpkgs?

Reply:
Merging it is proving rather ugly, since the stuff I can pass in at the top level isn't what I get inside all-packages.nix anymore. @Ericson2314 any ideas? I want to be able to put `@args` by the argument list at the top of all-packages.nix and then `import pkgs.path (args // overlay-junk)` but they don't match up.

Reply:
`all-packages` is just for a single stage of nixpkgs and therefore shouldn't import the whole thing. 

But.... it does still do that in a few more places, and for this purpose I defined  `nixpkgsFun` and passed it in. You might still clobber other overlays, but that can be dealt with in time.

Reply:
`all-packages` is just for a single stage of nixpkgs and therefore shouldn't import the whole thing. 

But.... it does still do that in a few more places, and for this purpose I defined  `nixpkgsFun` and passed it in. You might still clobber other overlays, but that can be dealt with in time.

Reply:
If nothing in `all-packages.nix` is going to *use* `headless`, I don't think we should put it in there. We should have a different file for "example nixpkgs instantations"---`release-lib.nix` and `release-cross.nix` could use it, for example. `lib/platforms/examples.nix` is somewhat similar.

Reply:
@Ericson2314 where would you put it? I'm looking for an easy way for people to interface with it, and the top-level all-packages.nix is generally the entry point into everything. We could change that, but I don't know what I'd change it to.

Reply:
I agree with @Ericson2314 that `all-packages.nix` is not a good place to put this. Historically we've put all sorts of non-package stuff in `all-packages.nix`, like `lib`, various functions (like `fetchurl`), etc etc. Ideally, only stuff intended to be usable by `nix-env` should be in `all-packages.nix`.

Also, if I read this correctly, it will cause Hydra to build thousands of extra packages, which is not a good idea.

Reply:
Adding another class of things (namely overlays) to `all-packages.nix` is really an RFC topic.

Reply:
```
lib.lowPrioSet (recurseIntoAttrs
```
Won't this be confusing for users of `nix-env`? Attribute names will be different but package names identical. Perhaps its better to remove `recurseIntoAttrs`?

Reply:
@FRidh the point of `recurseIntoAttrs` was to get Hydra to build it.

@edolstra yes, I don't know how many extra builds this will cause, but you're the one who added the no-x-libs.nix package set I think. Did you intend for anyone using minimal.nix to just recompile everything? It doesn't seem like a great user experience. I'd at least want to pick out some of the more common server packages and build those on Hydra, right?

Reply:
@copumpkin: it should work to use `recurseForRelease` instead of  `recurseForDerivations`.  (unused yet)

Reply:
Ooh, I didn't know about that one! I guess I'll put together an RFC though.

Reply:
@copumpkin A different file.

Reply:
But then usability is shot, right? I have someone on twitter saying "oh awesome, I can use headless packages? How do I just switch to using those?" I want it to be easy to get headless packages that you don't have to compile by hand.

I also don't see (outside of some notional "purity" argument) a fundamental reason to leave it out of all-packages. If our entry-point weren't all-packages.nix, in an ideal world, I wouldn't put it there, but that's how everyone gets into nixpkgs nowadays and I don't want it to be painful to get things like this, just because it isn't strictly a "package". There's so much crap in all-packages.nix today and drawing the line here without a pleasant alternative feels very theoretical at the cost of usability.

Reply:
It feels like perhaps something between impure.nix and all-packages.nix could include both this and all-packages.nix, as well as other things like "NixOS VM images" and other stuff that is painful to get at today.

Reply:
@copumpkin elaborating and improving on what I said on IRC.

1. There should be a canonical location to import all such overlays (and maybe other non-pkg-set-depending expressions). That way something like `(import <nixpkgs/pkgs/overlays>).fooBar` work and is cheap.

2. We can, for discoverability, make that (i.e `pkgs/overlays/default.nix` or whatever ends up being) accessible from `import <nixpkgs> {}`. `pkgs/top-level/default.nix` is between `pkgs/top-level/impure.nix` and `pkgs/top-level/stage.nix` as you and I both like.